odxtools/nameditemlist.py

```diff
@@ -1,10 +1,51 @@
 # SPDX-License-Identifier: MIT
 import warnings
 from keyword import iskeyword
-from typing import (Callable, Generic, Iterable, List, Optional, Tuple, TypeVar, Union, cast,
-                    overload)
+from typing import (Callable, Generic, Iterable, List, Optional, Protocol, Tuple, TypeVar, Union,
+                    cast, overload, runtime_checkable)
 
-T = TypeVar("T")
+from .exceptions import odxraise
+
+
+@runtime_checkable
+class OdxNamed(Protocol):
+
+    @property
+    def short_name(self) -> str:
+        pass
+
+
+T = TypeVar("T", bound=OdxNamed)
+
+
+def short_name_as_key(obj: OdxNamed) -> str:
+    """Retrieve an object's `short_name` attribute into a valid python identifier.
+
+    Although short names are almost identical to python identifiers,
+    their first character is allowed to be a number. This method
+    prepends an underscore to such such shortnames.
+    """
+    if not isinstance(obj, OdxNamed):
+        odxraise()
+    sn = obj.short_name
+    if not isinstance(sn, str):
+        odxraise()
+
+    if not sn.isidentifier():
+        warnings.warn(("For NamedItemList objects to work properly, all "
+                       "item names must be valid python identifiers."
+                       f"Encountered name '{sn}' which is not an "
+                       "identifier!"))
+
+    if sn[0].isdigit():
+        return f"_{sn}"
+
+    # make sure that the name of the item in question is not a
+    # python keyword (this would lead to syntax errors)
+    if iskeyword(sn):
+        return f"{sn}_"
+
+    return sn
 
 
 class NamedItemList(Generic[T]):
```

Comment:
IMO, it would be better to leave `NamedItemList` as is and derives a `ShortNameItemList` which  hard-codes the usage of `.short_name` as item names. possibly, I'm missing an issue with this, though?

Reply:
unless there is another sub class of NamedItemList other than ShortNameItemList, adding ShortNameItemList would be simply over engineering

Reply:
I would normally agree, but I'm thinking about using this class for other projects such as a residual bus simulator based on [cantools](https://github.com/cantools/cantools) where I need a different item name function.

(possibly, it is worth thinking about factoring this class out into a completely separate module...)

Reply:
NamedItemList have a base class ItemList, that allows overloading the key getter